src/detect-tls-certs.c

```diff
@@ -0,0 +1,263 @@
+ *
+ * \param de_ctx Pointer to the Detect Engine Context
+ * \param s      Pointer to the Signature to which the keyword belongs
+ * \param str    Should hold an empty string always
+ *
+ * \retval  0 On success
+ * \retval -1 On failure
+ */
+static int DetectTlsCertsSetup(DetectEngineCtx *de_ctx, Signature *s,
+                               const char *str)
+{
+    if (DetectBufferSetActiveList(s, g_tls_certs_buffer_id) < 0)
+        return -1;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_TLS) < 0)
+        return -1;
+
+    return 0;
+}
+
+static InspectionBuffer *TlsCertsGetData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *f,
+	struct TlsCertsGetDataArgs *cbdata, int list_id)
+{
+    SCEnter();
+
+    InspectionBufferMultipleForList *fb = InspectionBufferGetMulti(det_ctx, list_id);
+    InspectionBuffer *buffer = InspectionBufferMultipleForListGet(fb, cbdata->local_id);
+    if (buffer == NULL)
+        return NULL;
+
+    const SSLState *ssl_state = (SSLState *)f->alstate;
+
+    if (TAILQ_EMPTY(&ssl_state->server_connp.certs)) {
+        return NULL;
+    }
+
+    if (cbdata->cert == NULL) {
+        cbdata->cert = TAILQ_FIRST(&ssl_state->server_connp.certs);
+    } else {
+	cbdata->cert = TAILQ_NEXT(cbdata->cert, next);
+    }
+
+    if (cbdata->cert == NULL) {
+        return NULL;
+    }
+
+    InspectionBufferSetup(buffer, cbdata->cert->cert_data,
+		          cbdata->cert->cert_len);
+    InspectionBufferApplyTransforms(buffer, transforms);
+
+    SCReturnPtr(buffer, "InspectionBuffer");
+}
+
+static int DetectEngineInspectTlsCerts(DetectEngineCtx *de_ctx,
+	DetectEngineThreadCtx *det_ctx,
+	const DetectEngineAppInspectionEngine *engine,
+        const Signature *s, Flow *f, uint8_t flags, void *alstate, void *txv,
+	uint64_t tx_id)
+{
+    int local_id = 0;
+
+    const DetectEngineTransforms *transforms = NULL;
+    if (!engine->mpm) {
+        transforms = engine->v2.transforms;
+    }
+
+    struct TlsCertsGetDataArgs cbdata = { local_id, NULL };
+
+    while (1)
+    {
+        InspectionBuffer *buffer = TlsCertsGetData(det_ctx, transforms, f,
+			                           &cbdata, engine->sm_list);
+        if (buffer == NULL || buffer->inspect == NULL)
+            break;
+
+        det_ctx->buffer_offset = 0;
+        det_ctx->discontinue_matching = 0;
+        det_ctx->inspection_recursion_counter = 0;
+
+        const int match = DetectEngineContentInspection(de_ctx, det_ctx, s, engine->smd,
+                                              NULL, f, (uint8_t *)buffer->inspect,
+                                              buffer->inspect_len,
+                                              buffer->inspect_offset, DETECT_CI_FLAGS_SINGLE,
+                                              DETECT_ENGINE_CONTENT_INSPECTION_MODE_STATE);
+        if (match == 1) {
+            return DETECT_ENGINE_INSPECT_SIG_MATCH;
+        }
+
+	local_id++;
```

Comment:
on the 'local' local_id is incremented, not the one in the cbdata. So the same id is used for each iteration.

Reply:
I don't think you need to local var, you can just use and increment the one in cbdata directly I think

Reply:
Oh, thanks for pointing that out! I'll fix it right away!